### ISSUE TITLE
Am I doing something wrong? Current_Loss = 0.0000

### ISSUE THREAD

**Issue by chrisrapson** [GeorgeSeif/Semantic-Segmentation-Suite#89]
## Please specify the following information when submitting an issue:
- **What are your command line arguments?:**
`python train.py --dataset for_GeorgeSeif_training --crop_width 128 --crop_height 128 --h_flip false --v_flip false --model DeepLabV3 --frontend MobileNetV2`
- **Have you written any custom code?**:
I have made modifications to util.py 
1. to allow load_image to load grayscale masks
2. to give a warning, rather than an exception, if the image is smaller than the size to which it should be cropped in random_crop. (And then return the unaltered `image` and `label`)
- **What have you done to try and solve this issue?:**
Nothing, I am not sure if this is an issue or intended behaviour.
- **TensorFlow version?**:
1.10.1

### Describe the problem
The training routine seems to be running fine, but gives status updates that look like:
[2018-09-05 12:20:49] Epoch = 0 Count = 20 Current_Loss = 0.0000 Time = 3.27
[2018-09-05 12:20:52] Epoch = 0 Count = 40 Current_Loss = 0.0000 Time = 2.86
[2018-09-05 12:20:54] Epoch = 0 Count = 60 Current_Loss = 0.0000 Time = 2.92
[2018-09-05 12:20:58] Epoch = 0 Count = 80 Current_Loss = 0.0000 Time = 3.17
etc.

I expected that the Current_Loss would be non-zero, and that it would decrease as learning progresses. But perhaps I am misunderstanding something?

### Source code / logs
n/a

**Comment by GeorgeSeif** [comment]
Hi Chris, thanks for filling out the issue template in detail!

Mind posting an example image from your dataset? 

And by the way did you adjust the "class_dict.csv" to fit your dataset?

Thanks!

**Comment by chrisrapson** [comment]
No problem. Here is an example image and its mask. You might need to turn the brightness and contrast up to be able to see things in the mask, as the pixel values correspond to the label indices 1,2,3,4. (0=background.)

And here is my class_dict.csv:
name,r,g,b
head_light_L,128,250,200
head_light_R,128,200,250
brake_light_L,256,150,50
brake_light_R,256,50,150

![aachen_000000_000019_bbox01_leftimg8bit](https://user-images.githubusercontent.com/4982550/45066533-f1429380-b112-11e8-93d3-4df7a89a54e0.png)

![aachen_000000_000019_bbox01_leftimg8bit](https://user-images.githubusercontent.com/4982550/45066541-f869a180-b112-11e8-9b8a-8390cb4d2d78.png)

**Comment by GeorgeSeif** [comment]
I do see a potential problem. The whole code is really designed for RGB, though you should be able to get it working with a few modifications for the way your labels are.

First off, in addition to the utils, you may also need to modify the helper functions which convert to one shot as well as in the train.py, the input tensor is set for 3 channels, you'll need to change that to 1.

Also, your class_dict.csv has colours, but I see your labels are probably 0 to 255! there's a mismatch there.

Honestly, the easiest thing to do would be to convert your training images to have some kind of colour code. It'll work the easiest with the least hacking. Remember, the colours are only for visualisation purposes :)

**Comment by chrisrapson** [comment]
That worked, I got the training to run right through with a mini set of 50 training images. Now I'm training on the full set of images. It doesn't seem to be very stable, based on the plots of Average Loss and Average Validation Accuracy. Current_Loss is still 0 most of the time:
[2018-09-07 08:25:43] Epoch = 233 Count = 1460 Current_Loss = 0.0000 Time = 2.71
[2018-09-07 08:25:46] Epoch = 233 Count = 1480 Current_Loss = 0.0000 Time = 2.77
[2018-09-07 08:25:49] Epoch = 233 Count = 1500 Current_Loss = 2389103.7500 Time = 3.09
[2018-09-07 08:25:52] Epoch = 233 Count = 1520 Current_Loss = 0.0000 Time = 3.24
[2018-09-07 08:25:55] Epoch = 233 Count = 1540 Current_Loss = 2398237.5000 Time = 2.82
[2018-09-07 08:25:58] Epoch = 233 Count = 1560 Current_Loss = 0.0000 Time = 2.93
[2018-09-07 08:26:01] Epoch = 233 Count = 1580 Current_Loss = 0.0000 Time = 2.54
[2018-09-07 08:26:04] Epoch = 233 Count = 1600 Current_Loss = 0.0000 Time = 3.17
[2018-09-07 08:26:07] Epoch = 233 Count = 1620 Current_Loss = 386212.1562 Time = 2.84
[2018-09-07 08:26:10] Epoch = 233 Count = 1640 Current_Loss = 0.0000 Time = 3.18
[2018-09-07 08:26:13] Epoch = 233 Count = 1660 Current_Loss = 0.0000 Time = 3.16
[2018-09-07 08:26:16] Epoch = 233 Count = 1680 Current_Loss = 14898.5459 Time = 2.98
[2018-09-07 08:26:19] Epoch = 233 Count = 1700 Current_Loss = 0.0000 Time = 3.27

I suspect that's a different problem entirely, but since those are the only symptoms I have, I'm not sure how to go about submitting it as an issue.

**Comment by GeorgeSeif** [comment]
Wait.... which method did you end up trying? :)

**Comment by chrisrapson** [comment]
I converted my masks to RGB with the colours taken from my class_dict.csv file.

I still have images which are smaller than 128x128, where I ignore the exception raised in `random_crop()` and instead return the image unmodified. I have rolled back all my other modifications. If necessary, I guess I could up-sample all my images so that they are larger than 128x128.

**Comment by GeorgeSeif** [comment]
Hey Chris, 

I think it's definitely a data / set up issue. For the sake of thoroughness I tested your exact same command line arguments, just with CamVid dataset. It trains fine even with the 128x128 crop. 

The loss switching from massive values to 0 also leads me to suspect the data / setup issue.

**Comment by chrisrapson** [comment]
Ok, thanks for helping me narrow it down. Training on the CamVid dataset works for me, so it must be something to do with my image and label files. Do you have any suggestions for common issues I can troubleshoot? Here's a couple of things I can rule out so far

1. folder structure matches CamVid (class_dict.csv, train, train_labels, val, val_labels, test, test_labels)
2. RGB images in all folders
3. file names match between image and label folders. After running for a while, checkpoints fill with attempted validations comparing *_gt.png with *_pred.png.
4. upsampling to higher than 128x128 resolution still gives mostly Loss=0.

Could it be because the objects I have labelled are mostly small, so that >90% of the images are background?

**Comment by GeorgeSeif** [comment]
Hmm so I don't think it's because of small objects. That wouldn't really cause the loss to swing like that form 0 to an extreme value.

Can I see an image of one of your labels now that it's in RGB? It should basically look similar to how the CamVid labels do (coloured and you can see each colour). 

If we think about it, a zero loss mean that the network predicted the **exact** ground truth labels which is of course ridiculous! Have you taken a look at the validation images in the checkpoint folder? If you don't have any, try taking a miny chunk of the training set for validation so you can see some of the images as you train. That'll help a ton! :)

**Comment by chrisrapson** [comment]
Sure, here are a couple of image/mask pairs from my training set.

![aachen_000004_000019_bbox11_leftimg8bit](https://user-images.githubusercontent.com/4982550/45531394-b68fd800-b843-11e8-99b6-8542b1d36121.png)
![aachen_000004_000019_bbox11_leftimg8bit](https://user-images.githubusercontent.com/4982550/45531400-c0194000-b843-11e8-9fef-b6e0f3e8e925.png)

![aachen_000006_000019_bbox10_leftimg8bit](https://user-images.githubusercontent.com/4982550/45531454-eccd5780-b843-11e8-953d-17967074857d.png)
![aachen_000006_000019_bbox10_leftimg8bit](https://user-images.githubusercontent.com/4982550/45531470-fc4ca080-b843-11e8-97a0-465e8db1586d.png)

![berkeley_002b485a-3f6603f2_bbox06_leftimg8bit](https://user-images.githubusercontent.com/4982550/45531511-24d49a80-b844-11e8-85dc-d641cbd948ba.png)
![berkeley_002b485a-3f6603f2_bbox06_leftimg8bit](https://user-images.githubusercontent.com/4982550/45531497-14bcbb00-b844-11e8-9238-86c750f92e56.png)



And some of the pred/gt pairs that this suite saved in the checkpoints folder:

![kitti_000205_bbox01_leftimg8bit_pred](https://user-images.githubusercontent.com/4982550/45531351-85170c80-b843-11e8-8797-e2034d94f987.png)
![kitti_000205_bbox01_leftimg8bit_gt](https://user-images.githubusercontent.com/4982550/45531353-89432a00-b843-11e8-8010-884f36fc550c.png)

![berkeley_00067cfb-caba8a02_bbox04_leftimg8bit_pred](https://user-images.githubusercontent.com/4982550/45531568-79781580-b844-11e8-98c8-a0b3b9b2ee72.png)
![berkeley_00067cfb-caba8a02_bbox04_leftimg8bit_gt](https://user-images.githubusercontent.com/4982550/45531565-77ae5200-b844-11e8-86d1-dc1cdac3c12f.png)


By the way, I owe you a big thank you for taking the time to help me out here! I really appreciate it!

**Comment by GeorgeSeif** [comment]
Hmm wow. So looks like your labels are correct.

Unfortunately I think the only thing left to do is to go nitty gritty and start printing things line by line to and source the problem. I'd try and go through the pipeline step by step. I.e print after the first step, see if everything's OK, then go to the second step, etc. Break it down systematically. That's all there really is left to do!

And no problemo :)

**Comment by chrisrapson** [comment]
Ok, I took a deep dive and spotted a few things.
1. My class_dict.csv listed some of the RGB values as 256. I actually set the mask RGB values from the class_dict.csv, so I never considered that there could be a mismatch. But inside the 8-bit PNGs, 256 is automatically converted to 255. Then inside `train.py`, it doesn't match. I changed the 256s in class_dict.csv to 255s.
2. I found out that my rescaling program had a bug, and hadn't rescaled all my images to larger than 128x128. I fixed this, but I don't know whether it made any difference.
3. The `pred` and `gt` images in the checkpoints folders never had any background. It predicted one of my four classes for every pixel. See the images above which are all blue & green, instead of black. Since most of my images are background (unlike cityscapes or CamVid), predicting no background gave terrible scores, and didn't really learn anything. Adding a line to class_dict.csv: `background,0,0,0` made a big difference.

Now I seem to have a different problem - the routine has learnt that it gets close to a perfect score if it predicts background for the whole image. I see that you have a `compute_class_weights` function, but it doesn't appear to be used anywhere. I think that might be exactly what I need. Do you already have a plan for how that function should be used, or should I go ahead and try to apply it myself? (I'd be happy to supply a pull request if it works.)

**Comment by GeorgeSeif** [comment]
Well glad to hear that it works much better now! At least it's learning. 

So the reason I removed the usage option of class weights was because I could never find an implementation of it that worked well. Every time I tried, whether I used the weights or not I got very similar / the same results in the end. Feel free to try it if you like though!

How is the training otherwise? Are you getting a decent accuracy?

**Comment by chrisrapson** [comment]
I would say the training is "working", in that the routine has learnt that it can score very highly by predicting background for the whole image. That is not really incorrect, but it's not the result I am looking for. I will try two ways to get around that
1. brute force - learn for a large number of epochs, and hope that it discovers that a small benefit can be obtained by predicting my labels correctly
2. see if using weights makes a difference in my case.

**Comment by chrisrapson** [comment]
Instead of me reinventing the wheel, do you have an example where you have used the class weights?

**Comment by GeorgeSeif** [comment]
Hmm I'd definitely at least try the brute force approach. Might workout quite well! If you want to try class weights, checkout commit `43561f7 `. You don't even necessarily have to compute the class weights; it might be better to just set them yourself since that function has never really worked out too well for me

I'm closing this issue since there isn't any bug, but I'll monitor the comments to help you out :)

**Comment by chrisrapson** [comment]
brute force is making progress. It's been churning away on a GeForce GTX 1080 for 3 days now. For the first day, the loss increased steadily up to 10^9. Until that point the validation images were mostly flat background. From that point on, it started predicting sprays of pixels, and then blobs of foreground objects in roughly the right place. The loss has now dropped to 10^7. Only around 30% of the objects are detected, about half of those are the right class, and the shapes are way off... I'll keep it running over the weekend and see where it ends up.

I'm having a look at that commit now, and seeing if I can get something similar running to speed things up a bit.

Did you see the pull requests I created? Is there anything helpful in there?

**Comment by GeorgeSeif** [comment]
Seems to be taking a long time. Are you running on GPU?

**Comment by chrisrapson** [comment]
I agree, I was hoping it would be quicker. Yes, I'm pretty sure I'm running on a GPU. Based on `nvidia-smi` reporting 4459MiB of GPU memory usage by python, which disappears when I stop the training. It ran 175,000 epochs over the weekend. Each epoch takes about 0.3s with a batch size of 8 (crop size 128x128). When I did some test runs on a computer with no GPU it was much slower.

Unfortunately, things haven't really improved much over those 175,000 epochs. Could it be that this problem is too difficult for MobileNet to give an accurate result? Based on the examples I'd seen with a wider variety of more complex objects, I actually thought this would be a breeze.

**Comment by GeorgeSeif** [comment]
Yeah could be a MobileNet thing. Plus with MobileNet I had trouble loading the weights from the file, seems like bad indexing on the part of whoever saved it so I don't believe it's loading on the weights for that model. Try another perhaps!

Also, why are you using such a small image size? I rarely see anyone work with something as small as 128x128! Could be messing with the results as well whether you're doing cropping or just downsizing

**Comment by chrisrapson** [comment]
This is only part of an algorithm that I want to run at 30fps. Decreasing the image size makes the network much smaller, so it runs faster.

I decided to try combining Left & Right tail lights into one class, and the same for head lights. That is working much, much better. I think that for some reason the 4 class problem was too complex. Perhaps I didn't have enough images, or it was the limitations of MobileNet.

Once I am happy with the results using 2 classes, will I be able to use your routine to re-introduce 4 classes and see if transfer learning gives good results? Or will I have to keep the number of classes fixed?

**Comment by GeorgeSeif** [comment]
Yes, reducing the number of classes always helps improve accuracy!

Hmm unfortunately I don't have such transfer learning implemented since your last layer will change

I think that your are right that class balancing might help due to how unbalanced your data is!, especially since your main concern is to segment the tail lights. For your previous comment, you are correct in that the line **should be** `losses = unweighted_loss * weights`. Apologies for the late reply on that one.

That being said, I mentioned before how I didn't like that class weights function. So you can always try playing around and setting manually! Just change the below line

`class_weights = utils.compute_class_weights(labels_dir=args.dataset + "/train_labels", label_values=label_values)`

To a manual setting:

`class_weights = [0.3, 0.1, 0.4, 0.2]` 

Or something along those lines :)
